include/generic_sort.h

```diff
@@ -110,11 +119,13 @@ class GenericSort : public AbsSort
   // A string representation of a sort
   std::string to_string() const override;
 
+  std::string get_base_name() const;
+
  protected:
   // internal function to compute
   // the string representation of a sort
   std::string compute_string() const;
-
+  std::string base_name;
```

Comment:
OK maybe now that the PR is very mature, it is time to get rid of this field :) Basically, you should only need it in `GenericDatatypeSort`, and there you have a `Datatype` that you can down-cast to `GenericDatatype`, which has a `GenericDatatypeDecl`, which has a name.

Let me know where it goes wrong.

Reply:
What about the compare() function for the generic_sorts? That still has a case where I need to compare sorts of sortkind DATATYPE. I struggled previously with trying to cast the object itself to access the generic datatype. Is doing so even possible?

Reply:
So this should be done not by casting actually, but similarly to here:
https://github.com/makaimann/smt-switch/blob/5c0c1a0f3ef1e4578f6a8dad066a155fa3eb15fd/include/generic_sort.h#L95

This function is declared for `GenericSort`, but by default throws an exception. Only in some of the inheritting classes, it is implemented, using fields that only these inheritting classes have.

Similarly, you can have a fuction called `getDatatype` or something like that. By default, it will throw an exception. But in `GenericDatatypeSort`, it will return the `Datatype` field. Then you can use this field to extract the name. Sounds good?

Reply:
I'm still not following. How will I be able to use a function that's implemented for GenericDatatypeSort in GenericSort?

Reply:
You will define it for `GenericSort` and declare it with an implementation that throws an exception.
But on runtime, it will use the implementation of `GenericDatatypeSort`. 

This should have the exact structure of `get_uninterpreted_param_sorts`.

Is that clearer?

Reply:
Yes that does make sense. I'm surprised that it will use the GenericDatatypeSort on runtime. Why is that?

Reply:
It should always choose the implementation according to the actual type of the object. And I believe we will always `compare` a `DatatypeGenericSort` to a `DatatypeGenericSort`.